spec/requests/uploads_controller_spec.rb

```diff
@@ -774,7 +778,577 @@ def upload_file(file, folder = "images")
+      end
+    end
+
+    context "when the store is not external" do
+      before do
+        sign_in(user)
+      end
+
+      it "returns 404" do
+        post "/uploads/batch-presign-multipart-parts.json", params: {
+          unique_identifier: external_upload_stub.unique_identifier,
+          part_numbers: [1, 2, 3]
+        }
+        expect(response.status).to eq(404)
+      end
+    end
+  end
+
+  describe "#complete_multipart" do
+    let(:upload_base_url) { "https://#{SiteSetting.s3_upload_bucket}.s3.#{SiteSetting.s3_region}.amazonaws.com" }
+    let(:mock_multipart_upload_id) { "ibZBv_75gd9r8lH_gqXatLdxMVpAlj6CFTR.OwyF3953YdwbcQnMA2BLGn8Lx12fQNICtMw5KyteFeHw.Sjng--" }
+    let!(:external_upload_stub) do
+      Fabricate(:image_external_upload_stub, created_by: user, multipart: true, external_upload_identifier: mock_multipart_upload_id)
+    end
+
+    context "when the store is external" do
+      before do
+        sign_in(user)
+        SiteSetting.enable_direct_s3_uploads = true
+        setup_s3
+      end
+
+      def stub_list_multipart_request
+        list_multipart_result = <<~BODY
+        <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n
+        <ListPartsResult>
+           <Bucket>s3-upload-bucket</Bucket>
+           <Key>#{external_upload_stub.key}</Key>
+           <UploadId>#{mock_multipart_upload_id}</UploadId>
+           <PartNumberMarker>0</PartNumberMarker>
+           <NextPartNumberMarker>0</NextPartNumberMarker>
+           <MaxParts>1</MaxParts>
+           <IsTruncated>false</IsTruncated>
+           <Part>
+              <ETag>test</ETag>
+              <LastModified>#{Time.zone.now}</LastModified>
+              <PartNumber>1</PartNumber>
+              <Size>#{5.megabytes}</Size>
+           </Part>
+           <Initiator>
+              <DisplayName>test-upload-user</DisplayName>
+              <ID>arn:aws:iam::123:user/test-upload-user</ID>
+           </Initiator>
+           <Owner>
+              <DisplayName></DisplayName>
+              <ID>12345</ID>
+           </Owner>
+           <StorageClass>STANDARD</StorageClass>
+        </ListPartsResult>
+        BODY
+        stub_request(:get, "#{upload_base_url}/#{external_upload_stub.key}?uploadId=#{mock_multipart_upload_id}").to_return({ status: 200, body: list_multipart_result })
+      end
+
+      it "errors if the correct params are not provided" do
+        post "/uploads/complete-multipart.json", params: {}
+        expect(response.status).to eq(400)
+      end
+
+      it "errors if the part_numbers do not contain numbers between 1 and 10000" do
+        stub_list_multipart_request
+        post "/uploads/complete-multipart.json", params: {
+          unique_identifier: external_upload_stub.unique_identifier,
+          parts: [{ part_number: -1, etag: "test1" }]
+        }
+        expect(response.status).to eq(400)
+        expect(response.body).to include("Part numbers should be a list of numbers")
```

Comment:
The error seems to be incorrect since there is a list of `parts` here but no list of `part numbers`.

Reply:
No this is correct, the part number is -1 which is outside of the 1->10000 range

Reply:
Sorry I meant the error description. The parameter takes in a list of `parts` hashes and the error description is saying it should be a list of numbers which doesn't exists.

Reply:
Ah I see, changed it to:

>         "Each part number should be between 1 and 10000"